doc/rsp.md

```diff
@@ -280,95 +315,168 @@ mapped into the 16-bit lanes. Signed opcodes (`LPV`, `SPV`) map the value to bit
-the destination lane `de` is computed from the lowest 3 bits.
-
-`vt_elem` is 5 bits long (range 0..31). `vt_elem(4)` must be zero. When
-`vt_elem(3)` is 1, `vt_elem(2..0)` is actually used as source lane `se`, as expected. When
-`vt_elem(3)` is 0, a hardware bug is triggered and portions of the lower bits of
-`vt_elem` are replaced with portion of the bits of `vd_elem` while computing `se`. Specifically, all
-bits in `vt_elem` from the topmost set bit and higher are replaced with the
-same-position bits in `vt_elem`. Notice that this behaviour is actually consistent
-with what happens when `vt_elem(3)` is 1, which means that there is no need to
-think of it as a special-case. Pseudo-code:
-
-    de(2..0) = vd_elem(2..0)
-    msb = highest_set_bit(vt_elem)
-    se(2..0) = vd_elem(2..msb) || vt_elem(msb-1..0)
-
-TODO: complete analysis for `vt_elem(4)` == 1.
+single lange of a single input register (`VT<vt_elem>`), and store the result into
+a single lane of a single output register (`VD<vd_elem>`). Only the lowest 3
+bits of `vt_elem` and `vd_elem` are used to compute the source lane `se` and destination
+lane `de`, respectively.
+
+For all single-lane instructions, `vt_elem` is further used as a "broadcast modifier",
+(as in other SIMD architectures) that modifies `vt`, duplicating some lanes and hiding
+others. All 4 bits of `vt_elem` are used for this purpose. Note that the broadcast
+modifier `vt_elem` affects the value of `VT` whenever it is used within an instruction,
+but these modifications are not written back - that is, from the point of view of other
+instructions the value of `VT` remains the same.
 
+| `element` | Lanes being accessed | Description |
+| --- | --- | --- |
+| 0 | 0,1,2,3,4,5,6,7 | Normal register access (no broadcast) |
+| 1 | 0,1,2,3,4,5,6,7 | Normal register access (no broadcast) |
+| 2 | 0,0,2,2,4,4,6,6 | Broadcast 4 of 8 lanes |
+| 3 | 1,1,3,3,5,5,7,7 | Broadcast 4 of 8 lanes |
+| 4 | 0,0,0,0,4,4,4,4 | Broadcast 2 of 8 lanes |
+| 5 | 1,1,1,1,5,5,5,5 | Broadcast 2 of 8 lanes |
+| 6 | 2,2,2,2,6,6,6,6 | Broadcast 2 of 8 lanes |
+| 7 | 3,3,3,3,7,7,7,7 | Broadcast 2 of 8 lanes |
+| 8 | 0,0,0,0,0,0,0,0 | Broadcast single lane |
+| 9 | 1,1,1,1,1,1,1,1 | Broadcast single lane |
+| 10 | 2,2,2,2,2,2,2,2 | Broadcast single lane |
+| 11 | 3,3,3,3,3,3,3,3 | Broadcast single lane |
+| 12 | 4,4,4,4,4,4,4,4 | Broadcast single lane |
+| 13 | 5,5,5,5,5,5,5,5 | Broadcast single lane |
+| 14 | 6,6,6,6,6,6,6,6 | Broadcast single lane |
+| 15 | 7,7,7,7,7,7,7,7 | Broadcast single lane |
 
 VMOV
 ----
-Copy the source lane into the destination lane:
+Copy a lane from `vt` to `vd`, after broadcast:
 
-    VMOV vd[de],vs[se]
+    VMOV vd[de],vt[de]
 
 Pseudo-code:
 
-    VD<de> = VS<se>
+    VD<de> = VT<de>
 
+As a side-effect, `ACCUM_LO` is loaded with `VT`. Note that the source and destination
+lanes are both `de`, and `vt_elem` is being used as a broadcast modifier.
```

Comment:
Are you sure that `de` is also used for the source register after broadcast? Again this doesn't match my tests. Can you provide a RSP source code snippet that shows this?

Reply:
I haven't run the test myself, but I based this description off of the screenshot in https://github.com/PeterLemon/N64/tree/master/RSPTest/CP2/LOADSTORE/TransposeMatrixVMOV. The VMOVs used to construct each row all have the same `se`, but different values are loaded into each lane. The pattern seems to match the behavior if the source lane is also `de`. (I am also assuming the screenshot is from real hardware)

Reply:
No, what you see in the upper half of that screenshot is the behavior of the hardware bug that I described in the text you removed. The code is here:

https://github.com/PeterLemon/N64/blob/63f09b404bef9c0bdba6be9efa87ce78401864d7/RSPTest/CP2/LOADSTORE/TransposeMatrixVMOV/RSPTransposeMatrixVMOV.asm#L948

Each VMOV has a fixed `se` per row, but the register changes so they're different values. `de` instead is < 8, so the hardware bug is triggered. This is why the bottom half of the screen (generated with `de` >= 8) shows a correct matrix transposition, while the upper half is wrong.

BTW I regenerated most of the screenshots in that repo using my own N64 and dumping the framebuffer content, so they're mostly pixel-perfect. My own emulator's testsuite compares the output pixel by pixel.

Reply:
You can also explain the bottom half of the screen showing a correct matrix transposition with the broadcast modifier, since modifiers >= 8 broadcast one lane to the entire register. The first two rows of the top half are the same because modifiers 0 and 1 are the same. The entire test case  follows the pattern of applying each broadcast modifiers to the input matrix, then taking the diagonal.

I think your hardware bug replacing bits of `se` and the broadcast modifier might be different ways of describing the same thing, but am not 100% sure of that.

Reply:
Ah OK now I do understand. Your interpretation is actually made of two different parts: 1) vt_elem is used only as a broadcast modifier, 2) source and destination lane are the same.

This actually makes perfect sense and explains what I wrote as a "hardware bug". Kudos!

I think this should just be part of VMOV though. It's true that it's technically impossible to discern whether it applies to other single-instruction as you said, but at the same time the behavior makes sense only because there's also the part where source and dest lane are forced to be the same. I'd rather keep this whole weird behavior confined within VMOV.

Also I'd leave the broadcast table explanation later in the document. If we want, in a followup PR we can rearrange the order of the section so that it appears before single-lane instructions, but I wouldn't want to explain that in the context of VMOV. I'd rather it be part of the general description of computational instructions, and refer to it from VMOV.